File: part_2/python_version/harvesine.py
import json
import math

def calculateHarv(pair, earthRadius):
    lat1 = pair['y0']
    lat2 = pair['y1']
    lon1 = pair['x0']
    lon2 = pair['x1']

    dLat = lat2 - lat1
    dLon = lon2 - lon1

    a = math.sin(dLat/2)**2 + math.cos(lat1)* math.cos(lat2) * math.sin(dLon/2)**2
    c = 2*math.asin(math.sqrt(a))
    return earthRadius * c

dictionary = {
    "pairs": []
}

f = open('result.json')

data = json.load(f)

sum = 0
count = 0
for pair in data['pairs']:
    sum += calculateHarv(pair, 6372.8)
    count += 1
f.close()

print("count: ", count)
print("sum: ", sum)

File: part_2/pkg/harvesine/harvesine.go
package harvesine

import "math"

//EarthRadius is 6372.8
func ReferenceHarvesine(x0, y0, x1, y1, EarthRadius float64) float64 {
    lat1 := y0
    lat2 := y1
    lon1 := x0
    lon2 := x1

    dLat := radiansFromDegrees(lat2 - lat1) 
    dLon := radiansFromDegrees(lon2 - lon1) 
    lat1 = radiansFromDegrees(lat1)
    lat2 = radiansFromDegrees(lat2)

    a := Square(math.Sin(dLat/2.0) + math.Cos(lat1) * math.Cos(lat2) * Square(math.Sin(dLon/2))) 
    c := 2.0*math.Asin(math.Sqrt(a))

    return EarthRadius * c
}

func Square(A float64) float64 {
    return A * A
}

func radiansFromDegrees(degrees float64) float64 {
    return (math.Pi / 180) * degrees
}
